core/src/main/java/com/yahoo/imapnio/async/response/ImapResponseMapper.java

```diff
@@ -443,18 +456,102 @@ private SearchResult parseToSearchResult(@Nonnull final IMAPResponse[] ir) throw
             for (final IMAPResponse sr : ir) {
                 // There *will* be one SEARCH response.
                 if (sr.keyEquals("SEARCH")) {
                     while ((num = sr.readLong()) != -1) {
-                        v.add(Long.valueOf(num));
+                        v.add(num);
+                    }
+                    if (sr.readByte() == '(') {
+                        final String s = sr.readAtom();
+                        if (s.equalsIgnoreCase("MODSEQ")) {
+                            modSeq = sr.readLong();
+                        }
+                    }
+                }
+            }
+
+            return new SearchResult(v, modSeq);
+        }
+
+        /**
+         * Parses the responses from Store command and UID Store command to a @{code StoreResult} object.
+         *
+         * @param ir the list of responses from Store or UID Store command, the input responses array should contain the tagged/final one
+         * @return StoreResult object constructed based on the given IMAPResponse array,
+         * @throws ImapAsyncClientException when tagged response is not OK and NO or given response length is 0
+         */
+        @Nonnull
+        private StoreResult parseToStoreResult(@Nonnull final IMAPResponse[] ir) throws ImapAsyncClientException, IOException, ProtocolException {
+            if (ir.length < 1) {
+                throw new ImapAsyncClientException(FailureType.INVALID_INPUT);
+            }
+            final Response taggedResponse = ir[ir.length - 1];
+            taggedResponse.skipSpaces();
+            final byte bracket = taggedResponse.peekByte();
+            if (!taggedResponse.isOK() && bracket != (byte) L_BRACKET) { // BAD or NO without MODIFIED
+                throw new ImapAsyncClientException(FailureType.INVALID_INPUT);
+            }
+            final List<IMAPResponse> imapResponses = new ArrayList<>(); // will always return a non-null array
+
+            MessageNumberSet[] modifiedMsgsets = null; // only one response will contain modified numbers
+            Long highestModSeq = null;
+
+            for (final IMAPResponse sr: ir) {
+                if (sr.keyEquals("FETCH")) {
+                    imapResponses.add(new FetchResponse(sr));
+                    continue;
+                }
+
+                sr.skipSpaces();
+                if (sr.readByte() == (byte) L_BRACKET) { // HIGHESTMODSEQ or MODIFIED
+                    final String responseCode = sr.readAtom();
+                    if (sr.isOK() && responseCode.equalsIgnoreCase("HIGHESTMODSEQ")) {
+                        highestModSeq = sr.readLong();
+                    } else if (responseCode.equalsIgnoreCase("MODIFIED")) {
+                        modifiedMsgsets = MessageNumberSet.buildMessageNumberSets(sr.readAtom());
+                    } else { // OK or NO with '[' but response code is not recognized.
+                        throw new ImapAsyncClientException(FailureType.INVALID_INPUT);
```

Comment:
You should not be throwing exception at this point.
Send the raw response to caller. they can take a call.